lib/src/native/realm_core.dart

```diff
@@ -2004,53 +2004,54 @@ class _RealmCore {
     }
   }
 
-  static String? _appDir;
-
-  String _getAppDirectory() {
-    if (!isFlutterPlatform) {
-      return path.basenameWithoutExtension(File.fromUri(Platform.script).absolute.path);
-    }
-
-    if (Platform.isWindows || Platform.isLinux || Platform.isMacOS) {
-      const String libName = 'realm_plugin';
-      String binaryExt = Platform.isWindows
-          ? ".dll"
-          : Platform.isMacOS
-              ? ".dylib"
-              : ".so";
-      String binaryNamePrefix = Platform.isWindows ? "" : "lib";
-      final realmPluginLib =
-          Platform.isMacOS == false ? DynamicLibrary.open("$binaryNamePrefix$libName$binaryExt") : DynamicLibrary.open('realm.framework/realm');
-      final getAppDirFunc = realmPluginLib.lookupFunction<Pointer<Int8> Function(), Pointer<Int8> Function()>("realm_dart_get_app_directory");
-      final dirNamePtr = getAppDirFunc();
-      if (dirNamePtr == nullptr) {
-        return "";
-      }
+  String _getAppDirectoryFromPlugin() {
+    assert(isFlutterPlatform);
 
-      final dirName = Platform.isWindows ? dirNamePtr.cast<Utf16>().toDartString() : dirNamePtr.cast<Utf8>().toDartString();
-      return dirName;
-    }
+    String plugin = Platform.isWindows
+        ? 'realm_plugin.dll'
+        : Platform.isMacOS
+            ? 'realm.framework/realm' // use catalyst
+            : Platform.isLinux
+                ? "librealm_plugin.so"
+                : throw UnsupportedError("Platform ${Platform.operatingSystem} is not supported");
+
+    final pluginLib = DynamicLibrary.open(plugin);
+    final getAppDirFunc = pluginLib.lookupFunction<Pointer<Int8> Function(), Pointer<Int8> Function()>("realm_dart_get_app_directory");
+    final dirNamePtr = getAppDirFunc();
+    final dirName = Platform.isWindows ? dirNamePtr.cast<Utf16>().toDartString() : dirNamePtr.cast<Utf8>().toDartString();
 
-    return "";
+    return dirName;
   }
 
   String getAppDirectory() {
-    if (!isFlutterPlatform) {
-      return Directory.current.absolute.path;
+    if (!isFlutterPlatform || Platform.environment.containsKey('FLUTTER_TEST')) {
+      // return path.basenameWithoutExtension(File.fromUri(Platform.script).absolute.path);
+      return Directory.current.absolute.path; // dart or flutter test
     }
 
-    _appDir ??= _getAppDirectory();
+    // Flutter from here on..
 
     if (Platform.isAndroid || Platform.isIOS) {
-      return path.join(getFilesPath(), _appDir);
-    } else if (Platform.isWindows) {
-      return _appDir ?? Directory.current.absolute.path;
-    } else if (Platform.isLinux) {
-      String appSupportDir =
-          PlatformEx.fromEnvironment("XDG_DATA_HOME", defaultValue: PlatformEx.fromEnvironment("HOME", defaultValue: Directory.current.absolute.path));
-      return path.join(appSupportDir, ".local/share", _appDir);
-    } else if (Platform.isMacOS) {
-      return _appDir ?? Directory.current.absolute.path;
+      return getFilesPath();
```

Comment:
An an Android and iOS app already uses a directory local to the app. No need to make a subdirectory in it.. but you are are right this can break stuff for existing clients. Should we mention it under breaking changes, or revert a live with the odd extra directory level?

Reply:
This diff looks odd..

Reply:
`_appDir`  was always empty for iOS and Android the way it was calculated before. Files created on Android or iOS are local to the app by default.

Reply:
Could you double check that. Why would it be empty for iOS and Android. It because of these platforms that `realm_dart_get_app_directory` was introduced.

Reply:
There is a difference between:

- realm_dart_get_app_directory, and
- realm_dart_get_files_path

The first is implemented for linux, macos, and windows, the later for android and ios.